test/integration/puppetlabs/services/jruby/jruby_locking_test.clj

```diff
@@ -89,24 +98,35 @@
             (is (= [:instance-requested :instance-borrowed :instance-returned
                     :lock-requested :lock-acquired :lock-released] @events))))))))
 
-(deftest ^:integration borrow-and-return-affect-read-lock-test
-  (tk-bootstrap/with-app-with-config
-    app
-    [jruby-service/jruby-puppet-pooled-service
-     profiler/puppet-profiler-service]
-    (jruby-service-test-config 1)
-    (let [jruby-service (tk-app/get-service app :JRubyPuppetService)
-          lock (jruby-testutils/get-lock-from-app app)]
-
-      (is (= 0 (.getReadLockCount lock)))
-      (is (not (.isWriteLocked lock)))
-
-      (let [instance (jruby-protocol/borrow-instance jruby-service :test-borrow-and-read-lock)]
-        (testing "borrow-instance acquires a read lock and does not affect write lock"
-          (is (= 1 (.getReadLockCount lock)))
-          (is (not (.isWriteLocked lock))))
-
-        (testing "return-instance releases a read lock and does not affect write lock"
-          (jruby-protocol/return-instance jruby-service instance :test-return-and-read-lock)
-          (is (= 0 (.getReadLockCount lock)))
-          (is (not (.isWriteLocked lock))))))))
+(deftest ^:integration with-lock-and-borrow-contention-test
+  (testing "contention for instances with borrows and locking handled properly"
+    (tk-bootstrap/with-app-with-config
+     app
+     [jruby-service/jruby-puppet-pooled-service
+      profiler/puppet-profiler-service]
+     (jruby-service-test-config 2)
+     (jruby-testutils/wait-for-jrubies app)
+     (let [jruby-service (tk-app/get-service app :JRubyPuppetService)]
+       (let [instance (jruby-protocol/borrow-instance
+                       jruby-service
+                       :with-lock-and-borrow-contention-test)]
+         (let [lock-thread-started? (promise)
+               unlock-thread? (promise)
+               lock-thread (future (jruby-service/with-lock
+                                    jruby-service
+                                    :with-lock-and-borrow-contention-test
+                                    (deliver lock-thread-started? true)
+                                    @unlock-thread?))]
+           (testing "lock not granted yet when instance still borrowed"
+             (Thread/sleep 500)
```

Comment:
Yeah, I think there are other places where this could apply but where we haven't done the `sleep`.  I'll just remove the line for the sake of keeping the tests more snappy.  Hopefully the test will get to the `with-lock` call on the `future` often enough before the assertion occurs that we'll get some value from the assertion.